### ISSUE TITLE
grow_heap() should immediately drop and run the deferred alloc actions

### ISSUE THREAD

**Issue by kevinaboos** [theseus-os/Theseus#358]
the `grow_heap()` function works, but they may cause allocation requests for specific allocations to fail. 

See log demonstrating the issue:
```
[W] kernel/page_allocator/src/lib.rs:611: Requesting 2 pages starting at Some(v0xFFFFFE800A420000)
[I] kernel/multiple_heaps/src/lib.rs:468: grow_heap:: Allocated a page to refill core heap 1 for size :144 at address: 0xFFFFFE800A420000
[W] kernel/multiple_heaps/src/lib.rs:472: new heap_end: 0xFFFFFE800A422000
[W] kernel/multiple_heaps/src/lib.rs:474: deferred alloc action: DeferredAllocAction { free1: Chunk { pages: Page(v0xFFFFFE800A420000)..=Page(v0xFFFFFE800A41F000) }, free2: Chunk { pages: Page(v0xFFFFFE800A422000)..=Page(v0xFFFFFEFFFFFFF000) } }
[W] kernel/page_allocator/src/lib.rs:611: Requesting 2 pages starting at Some(v0xFFFFFE800A422000)
[W] kernel/page_allocator/src/lib.rs:387: For Pages Page(v0xFFFFFE800A422000) to Page(v0xFFFFFE800A423000), looking beyond chunk Chunk { pages: Page(v0x7B25000)..=Page(v0xFFFFFE7FFFFFF000) }
[W] kernel/page_allocator/src/lib.rs:389:      next chunk was chunk Some(Chunk { pages: Page(v0xFFFFFF8000000000)..=Page(v0xFFFFFFFF8009F000) })
```

### Solution

The fix is easy enough, we just need to drop the `DeferredAllocAction` object returned by `create_heap_mapping()` (for example, in Line 466 of  `multiple_heaps/src/lib.rs`) immediately instead of saving it. 
The current way that we save it til the end results in the subsequent contiguous chunk (above, the `free2` chunk from `Page(v0xFFFFFE800A422000) ..= Page(v0xFFFFFEFFFFFFF000)`) not being available until `action` is dropped. 

Alternatively, or additionally, we could allocate all N new heap chunks all at once, rather than iterating over `HEAP_GROWTH_AMOUNT`. The only downside is that it may fail more easily since we'd be asking for more contiguous pages, but for the heap this is not a real problem since the heap virtual addresses are already all reserved. 

Finally, we should tackle the to-do item in the page allocator that actually re-uses a removed chunk instead of always creating a new one no matter what.

**Comment by kevinaboos** [comment]
> The fix is easy enough, we just need to drop the DeferredAllocAction object returned by create_heap_mapping() (for example, in Line 466 of multiple_heaps/src/lib.rs) immediately instead of saving it.

Confirmed that this solution works.

**Comment by kevinaboos** [comment]
Closed by 11ea84f
